apps/heft/src/plugins/CopyFilesPlugin.ts

```diff
@@ -1,184 +1,96 @@
-      linkedFileCount
-    };
-  }
+  const { copiedFileCount, linkedFileCount } = await _copyFilesInnerAsync(copyDescriptors, logger);
+  logger.terminal.writeLine(
+    `Copied ${copiedFileCount} file${copiedFileCount === 1 ? '' : 's'} and ` +
+      `linked ${linkedFileCount} file${linkedFileCount === 1 ? '' : 's'}`
+  );
+
+  // TODO: Handle watch mode
+  // Then enter watch mode if requested
+  // if (options.watchMode) {
+  //   HeftAsync.runWatcherWithErrorHandling(async () => await this._runWatchAsync(options), logger);
+  // }
+}
 
-  private async _getCopyFileDescriptorsAsync(
-    buildFolder: string,
-    copyConfigurations: IResolvedDestinationCopyConfiguration[]
-  ): Promise<ICopyFileDescriptor[]> {
-    const processedCopyDescriptors: ICopyFileDescriptor[] = [];
+async function _getCopyFileDescriptorsAsync(
+  copyConfigurations: ICopyOperation[]
+): Promise<ICopyFileDescriptor[]> {
+  const processedCopyDescriptors: ICopyFileDescriptor[] = [];
 
-    // Create a map to deduplicate and prevent double-writes
-    // resolvedDestinationFilePath -> descriptor
-    const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
+  // Create a map to deduplicate and prevent double-writes
+  // resolvedDestinationFilePath -> descriptor
+  const destinationCopyDescriptors: Map<string, ICopyFileDescriptor> = new Map();
 
-    for (const copyConfiguration of copyConfigurations) {
-      // Resolve the source folder path which is where the glob will be run from
-      const resolvedSourceFolderPath: string = path.resolve(buildFolder, copyConfiguration.sourceFolder);
-      const sourceFileRelativePaths: Set<string> = new Set<string>(
-        await glob(this._getIncludedGlobPatterns(copyConfiguration), {
-          cwd: resolvedSourceFolderPath,
-          ignore: copyConfiguration.excludeGlobs,
-          dot: true,
-          onlyFiles: true
-        })
-      );
+  await Async.forEachAsync(
+    copyConfigurations,
+    async (copyConfiguration: ICopyOperation) => {
+      const sourceFileRelativePaths: Set<string> = await getRelativeFilePathsAsync(copyConfiguration);
 
       // Dedupe and throw if a double-write is detected
-      for (const destinationFolderPath of copyConfiguration.resolvedDestinationFolderPaths) {
+      for (const destinationFolderPath of copyConfiguration.destinationFolders) {
         for (const sourceFileRelativePath of sourceFileRelativePaths) {
           // Only include the relative path from the sourceFolder if flatten is false
-          const resolvedSourceFilePath: string = path.join(resolvedSourceFolderPath, sourceFileRelativePath);
+          const resolvedSourceFilePath: string = path.join(
+            copyConfiguration.sourceFolder,
+            sourceFileRelativePath
+          );
           const resolvedDestinationFilePath: string = path.resolve(
             destinationFolderPath,
             copyConfiguration.flatten ? '.' : path.dirname(sourceFileRelativePath),
```

Comment:
```suggestion
            copyConfiguration.flatten ? path.basename(sourceFileRelativePath) : sourceFileRelativePath
```
Then skip the next line as well. No point paying for split + join of the same path segments.